pallets/nodle-staking/src/lib.rs

```diff
@@ -0,0 +1,2311 @@
+            self.state = ValidatorStatus::Idle;
+        }
+        pub fn go_online(&mut self) {
+            self.state = ValidatorStatus::Active;
+        }
+        pub fn leave_validators_pool(&mut self, round: SessionIndex) {
+            self.state = ValidatorStatus::Leaving(round);
+        }
+    }
+
+    impl<AccountId, Balance> Validator<AccountId, Balance>
+    where
+        Balance: AtLeast32BitUnsigned + Saturating + Copy,
+    {
+        /// Slash the validator for a given amount of balance. This can grow the value
+        /// of the slash in the case that the validator has less than `minimum_balance`
+        /// active funds. Returns the amount of funds actually slashed.
+        ///
+        /// Slashes from `active` funds first, and then `unlocking`, starting with the
+        /// chunks that are closest to unlocking.
+        pub(crate) fn slash(&mut self, mut value: Balance, minimum_balance: Balance) -> Balance {
+            let pre_total = self.total;
+            let total = &mut self.total;
+            let active = &mut self.bond;
+
+            let slash_out_of =
+                |total_remaining: &mut Balance, target: &mut Balance, value: &Balance| {
+                    let mut slash_from_target = (*value).min(*target);
+
+                    if !slash_from_target.is_zero() {
+                        *target -= slash_from_target;
+
+                        // Make sure not drop below ED
+                        if *target <= minimum_balance {
+                            let diff_val = minimum_balance.saturating_sub(*target);
+                            *target += diff_val;
+                            slash_from_target -= diff_val;
+                        }
+
+                        *total_remaining = total_remaining.saturating_sub(slash_from_target);
+                    }
+                };
+
+            slash_out_of(total, active, &mut value);
+
+            let i = self
+                .unlocking
+                .iter_mut()
+                .map(|chunk| {
+                    slash_out_of(total, &mut chunk.value, &mut value);
+                    chunk.value
+                })
+                .take_while(|value| value.is_zero()) // take all fully-consumed chunks out.
+                .count();
+
+            // kill all drained chunks.
+            let _ = self.unlocking.drain(..i);
+
+            pre_total.saturating_sub(*total)
+        }
+        pub fn consolidate_unlocked(&mut self, current_session: SessionIndex) -> Balance {
+            let mut total = self.total;
+            self.unlocking.retain(|&chunk| {
+                if chunk.session_idx > current_session {
+                    true
+                } else {
+                    total = total.saturating_sub(chunk.value);
+                    false
+                }
+            });
+            let unlocked_val = self.total.saturating_sub(total);
+            self.total = total;
+            return unlocked_val;
+        }
+    }
```

Comment:
Probably deserves its own docstring to explain what it does

Reply:
:+1:  Done